hollow/src/main/java/com/netflix/hollow/api/consumer/incubating/HollowConsumerV2.java

```diff
@@ -0,0 +1,515 @@
+             *    b.withAPI(Charlie.class);
+             *    Consumer<Bravo> c3 = b.build(); // BOO
+             *
+             * Switch to immutable builder? That would confuse/break a lot of users who thought
+             * this was a drop-in replacement.
+             *
+             * Require it on construction?
+             *
+             *    Builder<Alpha> b = new Builder<>(Alpha.class);
+             */
+            requireNonNull(apiClass, "apiClass cannot be null");
+            if (HollowAPI.class.equals(apiClass))
+                throw new IllegalArgumentException("must provide a subclass of HollowAPI; apiClassFeature().reset() to restore platform default");
+            @SuppressWarnings("unchecked")
+            Builder<A2> self = (Builder<A2>)this;
+            self.apiClass.set(apiClass);
+            self.cachedTypes.reset();
+            return self;
+        }
+
+        /**
+         * Provide the code generated API class that extends {@link HollowAPI} with one or more types
+         * cached for direct field reads.
+         *
+         * All {@HollowRecord} instances are created by one of two factories:
+         *
+         * <dl>
+         *     <dt>{@link com.netflix.hollow.api.objects.provider.HollowObjectFactoryProvider}</dt>
+         *     <dd>creates an instance of the corresponding {@code <Type>DelegateLookupImpl} (e.g.
+         *     {@code StringDelegateLookupImpl} or {@code MovieDelegateLookupImpl} for core
+         *     types or types in a generated client API respectively). Field accesses perform
+         *     a lookup into the underlying high-density cache</dd>
+         *
+         *     <dt>{@link com.netflix.hollow.api.objects.provider.HollowObjectCacheProvider}</dt>
+         *     <dd>instantiates and caches the corresponding {@code <Type>DelegateCachedImpl} from
+         *     the generated client API (e.g. {@code MovieDelegateCachedImpl}). For a given ordinal,
+         *     the same {@code HollowRecord} instance is returned assuming the ordinal hasn't been removed.
+         *     All of the type's fields are eagerly looked up from the high-density cache and stored as Java fields,
+         *     aking field access in tight loops or the hottest code paths more CPU efficient.</dd>
+         * </dl>
+         *
+         * Object caching should only be enabled for low cardinality, custom types in your data model.
+         *
+         * Use {@link #withAPI(Class)} to build a consumer with your custom client API and
+         * using the default high-density cache for all types.
+         * @param cachedType a type to enable object caching on
+         * @param additionalCachedTypes additional types to enable object caching on
+         * @return this builder
+         *
+         * @see <a href="https://hollow.how/advanced-topics/#caching">https://hollow.how/advanced-topics/#caching</a>
+         */
+        public synchronized <A2 extends A> Builder<A2> withAPI(Class<A2> apiClass,
```

Comment:
Trivia: this entire PR is a yak shave that started because @DanielThomas wants the ability to specify cached types in Cinder.

Reply:
I'm not a troublemaker I swear.